clang/lib/Sema/SemaBounds.cpp

```diff
@@ -112,6 +112,90 @@ class BoundsUtil {
 };
 }
 
+namespace {
+  class VariableUtil {
+    public:
+      // If E is a (possibly parenthesized or no-op cast) variable V,
+      // GetVariable returns V.  Otherwise, GetVariable returns nullptr.
+      static DeclRefExpr *GetVariable(Sema &SemaRef, Expr *E) {
+        if (!E)
+          return nullptr;
+        return dyn_cast<DeclRefExpr>(E->IgnoreParenNoopCasts(SemaRef.Context));
+      }
+
+      // SameVariable returns true if the expressions E1 and E2
+      // are the same (possibly parenthesized or no-op cast) variable.
+      static bool SameVariable(Sema &SemaRef, Expr *E1, Expr *E2) {
```

Comment:
If you look at all uses of SameVariable, I believe the first expression is always a DeclRef expression.  I think it would be clearer to change the type of that argument.   You may need specialized versions for lvalue vs. rvalue expressions.